runtime/grid_reduction.cu

```diff
@@ -667,4 +667,243 @@ __device__ void serialReductionStep(
+  // "participating" threads
+  const auto id_in_block_segment =
+      index_utils::maskedOffset<!X_THREAD, !Y_THREAD, !Z_THREAD>(
+          threadIdx, blockDim);
+
+  // index into iteration dim
+  const auto thread_offset =
+      index_utils::maskedOffset<X_THREAD, Y_THREAD, Z_THREAD>(
+          threadIdx, blockDim);
+
+  // Stride by the "non-participating" threads
+  const auto input_stride_for_thread_in_segment =
+      index_utils::maskedSize<!X_THREAD, !Y_THREAD, !Z_THREAD>(blockDim);
+
+  // T inp = init_val;
+  T inp[vec_size];
+#pragma unroll
+  for (int i = 0; i < vec_size; i++) {
+    inp[i] = init_val;
+  }
+  // Block stride across the reduction until we only have one value per thread
+  for (nvfuser_index_t reduction_i = id_in_block_segment;
+       reduction_i < grid_reduction_segment_size;
+       reduction_i += input_stride_for_thread_in_segment) {
+    // index to global memory
+    auto work_buf_offset =
+        reduction_i * (block_reduction_segment_size * grid_segment_size) +
+        block_reduction_segment_size * idx_in_grid_segment + thread_offset;
+    work_buf_offset *= vec_size;
+
+    // vectorized load from global memory
+    constexpr int total_bytes = vec_size * sizeof(T);
+    constexpr int n_loads = total_bytes <= 16 ? 1 : total_bytes / 16;
+    constexpr int n_elements = total_bytes <= 16 ? vec_size : 16 / sizeof(T);
+#pragma unroll
+    for (int i = 0; i < n_loads; i++) {
+      int i_offset = i * n_elements;
+      T in_reg[n_elements];
+      loadGlobalToLocal<T, n_elements, true, CacheOp::Global>(
+          &in_reg[0], const_cast<T*>(in + work_buf_offset + i_offset));
+#pragma unroll
+      for (int j = 0; j < n_elements; j++) {
+        reduction_op(inp[i_offset + j], in_reg[j]);
+      }
+    }
+  }
+
+  // Block reduce the per thread values into per "participating" thread values
+  // T inp_tmp = init_val;
+  T inp_tmp[vec_size];
+#pragma unroll
+  for (int i = 0; i < vec_size; i++) {
+    inp_tmp[i] = init_val;
+  }
+  blockIterGroupedReduce<!X_THREAD, !Y_THREAD, !Z_THREAD, Aligned, vec_size>(
+      inp_tmp, inp, reduction_op, shared_buf, true, init_val);
+  const bool should_write = (X_THREAD || threadIdx.x == 0) &&
+      (Y_THREAD || threadIdx.y == 0) && (Z_THREAD || threadIdx.z == 0);
+  if (should_write && write_pred) {
+#pragma unroll
+    for (int i = 0; i < vec_size; i++) {
+      reduction_op(out[i], inp_tmp[i]);
+    }
+  }
+}
+
+template <
+    bool X_BLOCK,
+    bool Y_BLOCK,
+    bool Z_BLOCK,
+    bool X_THREAD,
+    bool Y_THREAD,
+    bool Z_THREAD,
+    bool PERSISTENT_REDUCTION,
+    bool Aligned,
+    int vec_size,
+    typename T,
+    typename Func>
+__device__ void iterGroupedGridReduce(
+    T* out,
+    const T* inp_val,
+    Func reduction_op,
+    volatile T* work_buf,
+    int64_t* sync_flags,
+    T* shared_buf,
+    bool read_pred,
+    bool write_pred,
+    T init_val,
+    const nvfuser_index_t entrance_ind,
+    const nvfuser_index_t n_entrances) {
```

Comment:
nitpick: n_entrances is just not used at all here. Should we remove it?

More interestingly, n_entrances doesn't seem to be used in gridReduce neither... Is that obsolete code?

Reply:
fyi, wea re also not using `entrance_ind` here at all...

Reply:
Nice catch! I removed `entrance_ind` and `n_entrances` in grouped version. 
For the original gridReduce, I'll do another PR to clean `n_entrances`.